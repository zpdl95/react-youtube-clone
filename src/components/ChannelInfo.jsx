import React from 'react';
import { useQuery } from '@tanstack/react-query';
import { useYoutubeApi } from '../context/YoutubeApiContext';

export default function ChannelInfo({ id, title }) {
  const { youtube } = useYoutubeApi();
  const { data: url } = useQuery({
    queryKey: ['channel', id],
    queryFn: ({ queryKey }) => {
      const [_, id] = queryKey;

      return youtube.channelImgURL(id);
    },
    staleTime: 1000 * 60 * 10,
  });

  return (
    <div className='flex items-center py-2 gap-3'>
      {url && <img src={url} alt={title} className='w-10 h-10 rounded-full' />}

      <p className='text-lg font-medium'>{title}</p>
    </div>
  );
}
