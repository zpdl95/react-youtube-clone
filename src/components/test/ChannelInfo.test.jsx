import React from 'react';
import { render, screen } from '@testing-library/react';
import { Route } from 'react-router-dom';
import { vi } from 'vitest';
import { withAllContexts, withRouter } from '../../test/utils';
import ChannelInfo from '../ChannelInfo';

describe('ChannelInfo', () => {
  const fakeYoutube = { channelImgURL: vi.fn() };

  afterEach(() => fakeYoutube.channelImgURL.mockReset());

  it('ChannelInfo snapshot', async () => {
    fakeYoutube.channelImgURL.mockImplementation(() => 'url');

    // asFragment = 이 함수를 실행시키면 그 타이밍의 렌더링 컴포넌트를 리턴한다
    const { asFragment } = renderChannelInfo();

    await screen.findByRole('img');
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders without URL', () => {
    fakeYoutube.channelImgURL.mockImplementation(() => {
      throw new Error('error');
    });

    renderChannelInfo();

    expect(screen.queryByRole('img')).toBeNull();
  });

  it('renders with URL', async () => {
    fakeYoutube.channelImgURL.mockImplementation(() => 'url');

    renderChannelInfo();

    await screen.findByRole('img');
  });

  function renderChannelInfo() {
    return render(
      withAllContexts(
        withRouter(
          <Route path='/' element={<ChannelInfo id='id' title='title' />} />
        ),
        fakeYoutube
      )
    );
  }
});
