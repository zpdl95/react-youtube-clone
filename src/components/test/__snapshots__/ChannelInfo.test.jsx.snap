// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChannelInfo > ChannelInfo snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center py-2 gap-3"
  >
    <img
      alt="title"
      class="w-10 h-10 rounded-full"
      src="url"
    />
    <p
      class="text-lg font-medium"
    >
      title
    </p>
  </div>
</DocumentFragment>
`;
